mautic/mautic
# Prevent some tokenized URLs in emails from being converted to trackables

**Description**

Some tokens are getting converted to trackables such as the webview_url and unsubscribe_url tokens. The redirects are thus sending all clicks to the first sent lead specific URL.  This PR addresses this by filtering out those tokens plus adds an event so that other bundles/plugins can inject tokens/URLs that should not be tracked.

This fixes #1288.

**Testing**

1. Create a list email that include the {webview_url}, {unsubscribe_url}, and {pagelink=ID} tokens as links and also a generic URL.  
2. Send the list email to more than one lead.
3. In each email, all four links should have been converted to trackables.  If you click on the links for the {webview_url} from each of the emails, they'll all redirect to the unique URL for the FIRST lead.
4. Apply the PR, clear the stats from the email_stats table or just clone the email and resend.  This time the {webview_url} and {unsubscribe_url} tokens should not be converted to trackables and be unique per lead.

**Hot Fix**
For those looking for a quick fix to this, simply edit this line https://github.com/mautic/mautic/blob/41a1d847fe79993efac708c082db93e3de31f202/app/bundles/PageBundle/EventListener/BuilderSubscriber.php#L595 to be 

```
            if (!in_array($token, array('{webview_url}', '{unsubscribe_url}'))) {
                $foundLinks[$url] = $currentTokens[$token];
            }
```
